src/transforms/pipelines/config.rs

```diff
@@ -41,62 +42,115 @@ impl Clone for PipelineConfig {
+                inner: Box::new(self.clone()),
+            },
+        );
+        // TODO: actually call outputs fn
         result
             .outputs
-            .push((last_name, vec![Output::default(DataType::all())]));
-        // insert the filter if needed and return the next inputs
-        let mut next_inputs = if let Some(ref filter) = self.filter {
-            let mut conditions = IndexMap::new();
-            conditions.insert("success".to_string(), filter.to_owned());
-            let filter_name = name.join("filter");
-            result.inner.insert(
-                filter_name.clone(),
-                InnerTopologyTransform {
-                    inputs: inputs.to_vec(),
-                    inner: Box::new(RouteConfig::new(conditions)),
-                },
-            );
-            result.outputs.push((
-                filter_name.clone(),
-                vec![Output::default(DataType::all()).with_port(UNMATCHED_ROUTE)],
-            ));
-            vec![filter_name.port("success")]
+            .push((name.clone(), vec![Output::default(DataType::all())]));
+
+        Ok(Some(result))
+    }
+}
+
+#[derive(Clone)]
+struct Pipeline {
+    condition: Option<Condition>,
+    transforms: Vec<Box<dyn SyncTransform>>,
+}
+
+impl SyncTransform for Pipeline {
+    fn transform(&mut self, _event: Event, _output: &mut TransformOutputsBuf) {
+        unimplemented!()
+    }
+
+    fn transform_all(&mut self, events: EventArray, output: &mut TransformOutputsBuf) {
+        if rand::random::<u16>() < 10 {
+            tracing::info!(batch_size = events.len());
+        }
+
+        let mut filtered = TransformOutputsBuf::dummy();
+
+        if let Some(condition) = &self.condition {
+            for event in events.into_events() {
+                if condition.check(&event) {
+                    filtered.push(event);
+                } else {
+                    output.push(event);
+                }
+            }
         } else {
-            inputs.to_vec()
-        };
-        // compound like
-        for (index, transform) in self.transforms.iter().enumerate() {
-            let step_name = name.join(index);
-            result.inner.insert(
-                step_name.clone(),
-                InnerTopologyTransform {
-                    inputs: next_inputs,
-                    inner: transform.to_owned(),
-                },
-            );
-            next_inputs = vec![step_name.id().to_string()];
+            filtered
+                .primary_buffer
+                .as_mut()
+                .unwrap()
+                .extend(events.into_events());
+        }
+
+        let mut tmp_out = filtered;
+        let mut tmp_in = TransformOutputsBuf::dummy();
+
+        for transform in &mut self.transforms {
+            std::mem::swap(&mut tmp_out, &mut tmp_in);
+            for event in tmp_in.primary_buffer.as_mut().unwrap().drain() {
+                transform.transform(event, output);
```

Comment:
Aaand that's a logic bug. Will fix and hope the soaks are still sufficiently interesting.